### PR TITLE
[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89981
* #89980
* __->__ #89900
* #89899
* #89898

**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern.

### CODE REVIEW

File: torch/distributed/fsdp/_optim_utils.py [hunk 1]
```diff
@@ -1323,3 +1372,100 @@ def return_fn(fqn_to_param_info):
...
+            else:
+                assert curr_value == [] or curr_value == value
+                orig_state[state_name] = value
+
+    # Unflatten state values.
+    for state_name in orig_state.keys():
+        value = orig_state[state_name]
+        if not isinstance(value, list) or not torch.is_tensor(value[0]):
+            continue
+        value = torch.concat(value)[: flat_param._numels[param_idx]].reshape(
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1426]: Conceptual: In what case would `torch.concat(value)` have more numel than `flat_param._numels[param_idx]` (meaning that we are truncating)?

File: torch/distributed/fsdp/_optim_utils.py [hunk 2]
```diff
@@ -1295,6 +1342,8 @@ def _optim_state_dict(
 def _get_fqn_to_fsdp_param_info(
     model: nn.Module, dedup_shared_fqns: Set[str]
 ) -> Dict[str, FSDPParamInfo]:
+    """Construct the mapping from fqn to the corrsponding FSDPParamInfo."""
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1345]: nit:
```suggestion
    """Construct the mapping from fqn to the corresponding FSDPParamInfo."""
```

File: torch/distributed/fsdp/_optim_utils.py [hunk 3]
```diff
@@ -1161,7 +1141,63 @@ def _map_param_id_to_optim_keys(
...
 
-    return r0_param_id_to_optim_state_key, optim_state_key_to_param_id
+
+def _map_param_id_to_optim_keys(
+    optim_state_dict: Dict[str, Any],
+    group: Optional[dist.ProcessGroup],
+    param_id_to_param: List[nn.Parameter],
+    param_to_fqns: Dict[nn.Parameter, List[str]],
+    fqn_to_fsdp_param_info: Dict[str, FSDPParamInfo],
+    merge_key: bool = False,
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1151]: Do we need `merge_key=True` for `use_orig_params=True` to handle the fact that ranks who do not have any part of an original parameter will not have any optimizer state for that parameter?

Then, we also have `merge_key=False` to preserve the existing behavior for `use_orig_params=False`?

**Review by awgu — COMMENTED** [review]
Big thanks for working on this and getting through the crazy Cartesian product of code paths!

I made an initial pass and will continue to revisit.

File: torch/distributed/fsdp/_optim_utils.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:1426]: Excellent point. The padding shouldn't be visible here. However, I'm not sure if it is always true as we were discussing the cost of F.pad and may change the implementation. This is just a safety guard. If you believe padding will never be seen, I can remove it.

File: torch/distributed/fsdp/_optim_utils.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:1151]: That's correct.

File: torch/distributed/fsdp/_optim_utils.py [hunk 4]
```diff
@@ -1086,50 +1100,16 @@ def _is_zero_dim_tensor(x: Any) -> bool:
     return torch.is_tensor(x) and x.dim() == 0
 
 
-def _map_param_id_to_optim_keys(
-    optim_state_dict: Dict[str, Any],
-    group: Optional[dist.ProcessGroup],
+def _check_rank_missing_keys(
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1103]: nit:
```suggestion
def _check_missing_keys_on_rank(
```

File: torch/distributed/fsdp/_optim_utils.py [hunk 5]
```diff
@@ -1159,7 +1139,64 @@ def _map_param_id_to_optim_keys(
...
+def _map_param_id_to_optim_keys(
+    optim_state_dict: Dict[str, Any],
+    group: Optional[dist.ProcessGroup],
+    param_id_to_param: List[nn.Parameter],
+    param_to_fqns: Dict[nn.Parameter, List[str]],
+    fqn_to_fsdp_param_info: Dict[str, FSDPParamInfo],
+    merge_keys: bool = False,
+) -> Tuple[List[_OptimStateKey], Dict[_OptimStateKey, int]]:
+    """
+    Construct the local mapping between the ``_OptimStateKey`` and parameter IDs
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1152]: Thanks for adding this documentation!

File: torch/distributed/fsdp/_optim_utils.py [hunk 6]
```diff
@@ -1238,34 +1276,44 @@ def _optim_state_dict(
...
+        merge_keys=use_orig_params,
     )
 
     # Iterate in rank 0's flattened parameter ID order to ensure aligned
     # all-gathers across ranks
-    for optim_state_key in param_id_to_optim_state_key.values():
-        param_id = optim_state_key_to_param_id[optim_state_key]
+    for optim_state_key in all_optim_state_keys:
+        param_id = optim_state_key_to_param_id.get(optim_state_key, -1)
+        assert param_id >= 0 or (optim_state_key.is_fsdp_managed and use_orig_params)
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1292]: nit: Can we add a message to the `assert` explaining that for `use_orig_params=True`, some parameters may not be in the local shard, so they will be missing the optimizer state keys?

File: torch/distributed/fsdp/_optim_utils.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/_optim_utils.py:1431]: Ah good catch. Let us keep the trimming logic here. The safe guard is good.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -1344,12 +1348,18 @@ def full_optim_state_dict(
...
-        FullyShardedDataParallel._raise_on_use_orig_params_optim_checkpoint(model)
+        FullyShardedDataParallel._raise_on_use_orig_params_optim_checkpoint(
+            model, True, rank0_only
+        )
         FullyShardedDataParallel._warn_optim_input(optim_input)
         using_optim_input = FullyShardedDataParallel._is_using_optim_input(
             optim_input,
             optim,
         )
+        use_orig_params: bool = False
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1216]: cc: @rohan-varma We were talking about which FSDP constructor args we assume to be uniform. This is one place evidencing that `use_orig_params` must be uniform (even if we already intuitively thought that already).

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 8]
```diff
@@ -1652,6 +1676,69 @@ def rekey_optim_state_dict(
...
+        based on ``state_dict_type`` set by :meth:`set_state_dict_type` or
+        :meth:`state_dict_type`.
+
+        Args:
+            model (torch.nn.Module): Root module (which may or may not be a
+                :class:`FullyShardedDataParallel` instance) whose parameters
+                were passed into the optimizer ``optim``.
+            optim (torch.optim.Optimizer): Optimizer for ``model`` 's
+                parameters.
+                the default process group. (Default: ``None``)
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1700]: It looks like some part got cut off here for `group`?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 9]
```diff
@@ -1652,6 +1676,69 @@ def rekey_optim_state_dict(
...
+        to load the state_dict.
+
+        Args:
+            state_dict (Dict[str, Any]): The optimizer states to be loaded.
+            model (torch.nn.Module): Root module (which may or may not be a
+                :class:`FullyShardedDataParallel` instance) whose parameters
+                were passed into the optimizer ``optim``.
+            optim (torch.optim.Optimizer): Optimizer for ``model`` 's
+                parameters.
+            group (dist.ProcessGroup): Model's process group or ``None`` if using
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1734]: nit: For both here and `_optim_state_dict`, let us mention that this is the process group across which parameters are sharded since we have landed HSDP, which has both the intra-node sharding process group and the inter-node replicating process group.
```suggestion
            group (dist.ProcessGroup): Model's process group across which parameters are sharded or ``None`` if using
```

File: torch/distributed/fsdp/_optim_utils.py [hunk 10]
```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
...
+                torch.cuda.device_count(),
+                cast(dist.ProcessGroup, fsdp_state.process_group),
+            )
+        value = value.cpu()
+        orig_state[state_name] = value
+    return orig_state
+
+
+def _shard_orig_param_state(
+    fqn: str,
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1448]: nit: For this function and `_gather_orig_param_state`, can we specify if `fqn` is the local FQN (as stored in `flat_param._fqns`) or the global FQN (which may require prepending a prefix starting from the local FSDP root).

Ugh, we might need to clarify our terminology at some point and just write it down somewhere.

File: torch/distributed/fsdp/_optim_utils.py [hunk 11]
```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
...
+    # Gathering state from all ranks. This step may be slow. However,
+    # `state_dict()` is not in the critical path. We can fuse the communication
+    # if the performance becomes a problem.
+    state_objects = {
+        state_name: value for state_name, value in sorted_items(optim_state)
+    }
+    object_list: List[Dict[str, Any]] = [
+        {} for _ in range(cast(int, fsdp_state.world_size))
+    ]
+    dist.all_gather_object(object_list, state_objects)
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1410]: I forget; for the `use_orig_params=False` case, what collective do we use? `all_gather_into_tensor`?

**Review by awgu — APPROVED** [review]
This looks good to me! Hopefully, you can get another review from someone else as well, but it is not strictly necessary.

I left some minor comments.

**Review by rohan-varma — APPROVED** [review]
LGTM, will let @awgu to accept it as well, thanks!

File: torch/distributed/fsdp/_optim_utils.py (hunk 11 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:1410]: That's correct.

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 12]
```diff
@@ -749,32 +750,44 @@ def test_scatter_full_optim_state_dict_transformer(self) -> None:
...
+            use_optim_input=False,
+        )
+
+    @skip_if_lt_x_gpu(2)
+    def test_use_orig_params(self) -> None:
+        """Tests :meth:`optim_state_dict` for an FSDP-root nested model."""
+        self._test_load_optim_state(
+            _ModelClass.NESTED,
+            use_multiple_param_groups=False,
+            halve_world_size=False,
```
[the remark this answers is not in the document]
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_optim_state.py:785]: More tests will be introduced after we make `FSDP.optim_state_dict` support all use cases.

File: torch/distributed/fsdp/_optim_utils.py [hunk 13]
```diff
@@ -1159,7 +1139,64 @@ def _map_param_id_to_optim_keys(
...
+    for param_id, param in enumerate(param_id_to_param):
+        # Do not include parameters without state to avoid empty mappings
+        # just like in normal `torch.optim.Optimizer.state_dict()`
+        if param_id not in optim_state_dict["state"]:
+            continue
+        fqns = param_to_fqns[param]
+        is_fsdp_managed = isinstance(param, FlatParameter)
+        if is_fsdp_managed:
+            assert fqns[0] in fqn_to_fsdp_param_info
+        is_fsdp_managed = fqns[0] in fqn_to_fsdp_param_info
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:1170]: Yes, for `use_orig_params` case, this line is required.

File: torch/distributed/fsdp/_optim_utils.py [hunk 14]
```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
...
+        )
+        if shard_state:
+            value = _ext_chunk_tensor(
+                value,
+                cast(int, fsdp_state.rank),
+                cast(int, fsdp_state.world_size),
+                torch.cuda.device_count(),
+                cast(dist.ProcessGroup, fsdp_state.process_group),
+            )
+        value = value.cpu()
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:1442]: Yes, we do in the original code path. This is the inconsistency between `state_dict` and `optim_state_dict`. I had a PR to fix this. But it never land. Will introduce this later when I complete `optim_state_dict()`.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 15]
```diff
@@ -1652,6 +1676,69 @@ def rekey_optim_state_dict(
...
+        optim: torch.optim.Optimizer,
+        group: Optional[dist.ProcessGroup] = None,
+    ) -> Dict[str, Any]:
+        """
+        This API is still being developed, hence the `_` prefix. The comment
+        below is also not fully implemented yet. Do not use this API unless
+        you know why this API exists and how this API works.
+
+        Returns the optimizer state. The state will be sharded or consolidated
+        based on ``state_dict_type`` set by :meth:`set_state_dict_type` or
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1691]: We have not enforced this yet. But it will be enforced when `optim_state_dict` support all use cases.
